File: MiniDOMBase/DMD/Luoghi/Province/CProvincia.cs
using System;
using System.Collections;
using DMD;
using DMD.XML;
using DMD.Databases;
using DMD.Databases.Collections;
using static minidom.Sistema;
using static minidom.Anagrafica;
using minidom.repositories;
using System.Collections.Generic;


namespace minidom
{
    public partial class Anagrafica
    {

        /// <summary>
        /// Rappresenta una provincia
        /// </summary>
        [Serializable]
        public class CProvincia 
            : Luogo
        {
            private string m_Sigla;
            private string m_Regione;
            
            /// <summary>
            /// Costruttore
            /// </summary>
            public CProvincia()
            {
                this.m_Sigla = "";
                this.m_Regione = "";                 
            }

            /// <summary>
            /// Repository
            /// </summary>
            /// <returns></returns>
            public override CModulesClass GetModule()
            {
                return minidom.Anagrafica.Luoghi.Province; //.Module;
            }

            /// <summary>
            /// Restituisce o imposta la sigla della provincia
            /// </summary>
            public string Sigla
            {
                get
                {
                    return m_Sigla;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_Sigla;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Sigla = value;
                    DoChanged("Sigla", value, oldValue);
                }
            }

            /// <summary>
            /// Restituisce o imposta la regione
            /// </summary>
            public string Regione
            {
                get
                {
                    return m_Regione;
                }

                set
                {
                    value = DMD.Strings.Trim(value);
                    string oldValue = m_Regione;
                    if ((oldValue ?? "") == (value ?? ""))
                        return;
                    m_Regione = value;
                    DoChanged("Regione", value, oldValue);
                }
            }

            //protected internal override Databases.CDBConnection GetConnection()
            //{
            //    return Luoghi.Database;
            //}

            /// <summary>
            /// Discriminator
            /// </summary>
            /// <returns></returns>
            public override string GetTableName()
            {
                return "tbl_Luoghi_Province";
            }

            /// <summary>
            /// Carica dal db
            /// </summary>
            /// <param name="reader"></param>
            /// <returns></returns>
            protected override bool LoadFromRecordset(DBReader reader)
            {
                this.m_Sigla = reader.Read("Sigla", this.m_Sigla);
                this.m_Regione = reader.Read("Regione", this.m_Regione);
                return base.LoadFromRecordset(reader);
            }

            /// <summary>
            /// Salva nel db
            /// </summary>
            /// <param name="writer"></param>
            /// <returns></returns>
            protected override bool SaveToRecordset(DBWriter writer)
            {
                writer.Write("Sigla", m_Sigla);
                writer.Write("Regione", m_Regione);
                return base.SaveToRecordset(writer);
            }

            /// <summary>
            /// Prepara lo schema
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaFields(DBTable table)
            {
                base.PrepareDBSchemaFields(table);

                var c = table.Fields.Ensure("Sigla", typeof(string), 255);
                c = table.Fields.Ensure("Regione", typeof(string), 255);
            }

            /// <summary>
            /// Prepara i vincoli
            /// </summary>
            /// <param name="table"></param>
            protected override void PrepareDBSchemaConstraints(DBTable table)
            {
                base.PrepareDBSchemaConstraints(table);

                var c = table.Constraints.Ensure("idxSigla", new string[] { "Sigla", "Regione" }, DBFieldConstraintFlags.None);
            }

            /// <summary>
            /// Restituisce una stringa che rappresenta l'oggetto
            /// </summary>
            /// <returns></returns>
            public override string ToString()
            {
                return Nome;
            }

            /// <summary>
            /// Restituisce il codice hash dell'oggetto
            /// </summary>
            /// <returns></returns>
            public override int GetHashCode()
            {
                return base.GetHashCode();
            }

            /// <summary>
            /// Serializzazione xml
            /// </summary>
            /// <param name="writer"></param>
            protected override void XMLSerialize(XMLWriter writer)
            {
                writer.WriteAttribute("m_Sigla", m_Sigla);
                writer.WriteAttribute("m_Regione", m_Regione);
                base.XMLSerialize(writer);
            }

            /// <summary>
            /// Deserializzazione xml
            /// </summary>
            /// <param name="fieldName"></param>
            /// <param name="fieldValue"></param>
            protected override void SetFieldInternal(string fieldName, object fieldValue)
            {
                switch (fieldName ?? "")
                {
                    case "m_Sigla":
                        {
                            m_Sigla = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    case "m_Regione":
                        {
                            m_Regione = DMD.XML.Utils.Serializer.DeserializeString(fieldValue);
                            break;
                        }

                    default:
                        {
                            base.SetFieldInternal(fieldName, fieldValue);
                            break;
                        }
                }
            }

            /// <summary>
            /// Restituisce true se gli oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public override bool Equals(Luogo obj)
            {
                return (obj is CProvincia) && this.Equals((CProvincia)obj);
            }

            /// <summary>
            /// Restituisce true se gli oggetti sono uguali
            /// </summary>
            /// <param name="obj"></param>
            /// <returns></returns>
            public virtual bool Equals(CProvincia obj)
            {
                return base.Equals(obj)
                    && DMD.Strings.EQ(this.m_Sigla, obj.m_Sigla)
                    && DMD.Strings.EQ(this.m_Regione, obj.m_Regione)
                    ;
            }

        }
    }
}

File: MiniDOMBase/DMD/Luoghi/Province/CProvincia.vb
Imports minidom
Imports minidom.Sistema
Imports minidom.Databases
Imports minidom.Anagrafica

Partial Public Class Anagrafica

    <Serializable> _
    Public Class CProvincia
        Inherits Luogo

        Private m_Sigla As String
        Private m_Regione As String
        Private m_NumeroAbitanti As Integer

        Public Sub New()
            Me.m_Sigla = ""
            Me.m_Regione = ""
            Me.m_NumeroAbitanti = 0
        End Sub

        Public Overrides Function GetModule() As CModule
            Return Luoghi.Province.Module
        End Function

        Public Property NumeroAbitanti As Integer
            Get
                Return Me.m_NumeroAbitanti
            End Get
            Set(value As Integer)
                Dim oldValue As Integer = Me.m_NumeroAbitanti
                If (oldValue = value) Then Exit Property
                Me.m_NumeroAbitanti = value
                Me.DoChanged("NumeroAbitanti", value, oldValue)
            End Set
        End Property

        Public Property Sigla As String
            Get
                Return Me.m_Sigla
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_Sigla
                If (oldValue = value) Then Exit Property
                Me.m_Sigla = value
                Me.DoChanged("Sigla", value, oldValue)
            End Set
        End Property

        Public Property Regione As String
            Get
                Return Me.m_Regione
            End Get
            Set(value As String)
                value = Trim(value)
                Dim oldValue As String = Me.m_Regione
                If (oldValue = value) Then Exit Property
                Me.m_Regione = value
                Me.DoChanged("Regione", value, oldValue)
            End Set
        End Property

        Protected Friend Overrides Function GetConnection() As Databases.CDBConnection
            Return Anagrafica.Luoghi.Database
        End Function

        Public Overrides Function GetTableName() As String
            Return "tbl_Luoghi_Province"
        End Function

        Protected Overrides Function LoadFromRecordset(ByVal reader As DBReader) As Boolean
            Me.m_NumeroAbitanti = reader.Read("NumeroResidente", Me.m_NumeroAbitanti)
            Me.m_Sigla = reader.Read("Sigla", Me.m_Sigla)
            Me.m_Regione = reader.Read("Regione", Me.m_Regione)
            Return MyBase.LoadFromRecordset(reader)
        End Function

        Protected Overrides Function SaveToRecordset(ByVal writer As DBWriter) As Boolean
            writer.Write("NumeroResidente", Me.m_NumeroAbitanti)
            writer.Write("Sigla", Me.m_Sigla)
            writer.Write("Regione", Me.m_Regione)
            Return MyBase.SaveToRecordset(writer)
        End Function

        Public Overrides Function ToString() As String
            Return Me.Nome
        End Function

        Protected Overrides Sub XMLSerialize(ByVal writer As minidom.XML.XMLWriter)
            writer.WriteAttribute("m_NumeroAbitanti", Me.m_NumeroAbitanti)
            writer.WriteAttribute("m_Sigla", Me.m_Sigla)
            writer.WriteAttribute("m_Regione", Me.m_Regione)
            MyBase.XMLSerialize(writer)
        End Sub

        Protected Overrides Sub SetFieldInternal(ByVal fieldName As String, ByVal fieldValue As Object)
            Select Case (fieldName)
                Case "m_NumeroAbitanti" : m_NumeroAbitanti = XML.Utils.Serializer.DeserializeInteger(fieldValue)
                Case "m_Sigla" : m_Sigla = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case "m_Regione" : m_Regione = XML.Utils.Serializer.DeserializeString(fieldValue)
                Case Else : MyBase.SetFieldInternal(fieldName, fieldValue)
            End Select
        End Sub

        Protected Overrides Sub OnAfterSave(e As SystemEvent)
            MyBase.OnAfterSave(e)
            Luoghi.Province.UpdateCached(Me)
        End Sub
    End Class


End Class